src/session.c

```diff
@@ -0,0 +1,449 @@
+done:
+    if (ret == CKR_OK) {
+        session->pool->login_session = p11prov_session_ref(session);
+        LIST_INSERT(session->pool->busy_sessions, session);
+    } else {
+        p11prov_session_free(session);
+    }
+    OPENSSL_cleanse(cb_pin, cb_pin_len);
+    return ret;
+}
+
+CK_RV p11prov_get_session(P11PROV_CTX *provctx, CK_SLOT_ID *slotid,
+                          CK_SLOT_ID *next_slotid, P11PROV_URI *uri,
+                          OSSL_PASSPHRASE_CALLBACK *pw_cb, void *pw_cbarg,
+                          P11PROV_SESSION **session)
+{
+    P11PROV_SESSION *sess;
+    CK_SLOT_ID id = *slotid;
+    struct p11prov_slot *slots = NULL;
+    struct p11prov_slot *slot = NULL;
+    int nslots = 0;
+    int i;
+    CK_RV ret = CKR_CANCEL;
+    bool check_session = false;
+
+    nslots = p11prov_ctx_get_slots(provctx, &slots);
+
+    for (i = 0; i < nslots; i++) {
+        if (id != CK_UNAVAILABLE_INFORMATION && id != slots[i].id) {
+            continue;
+        }
+
+        /* ignore slots that are not initialized */
+        if ((slots[i].slot.flags & CKF_TOKEN_PRESENT) == 0) {
+            continue;
+        }
+        if ((slots[i].token.flags & CKF_TOKEN_INITIALIZED) == 0) {
+            continue;
+        }
+
+        id = slots[i].id;
+        slot = &slots[i];
+        ret = CKR_OK;
+
+        if (uri) {
+            CK_TOKEN_INFO token = slot->token;
+
+            /* skip slots that do not match */
+            ret = p11prov_uri_match_token(uri, &token);
+            if (ret == CKR_CANCEL) {
+                continue;
+            }
+            if (ret == CKR_OK && (token.flags & CKF_LOGIN_REQUIRED)) {
+                pthread_mutex_lock(&slot->sessions->lock);
+                ret = token_login(provctx, slot, uri, pw_cb, pw_cbarg);
+                pthread_mutex_unlock(&slot->sessions->lock);
+            }
+        }
+        break;
+    }
+
+    if (ret == CKR_OK) {
+        /* Found a slot, return it and the next slot to the caller for
+         * continuation if the current slot does not yield the desired
+         * results */
+        *slotid = id;
+        if (next_slotid) {
+            if (i + 1 < nslots) {
+                *next_slotid = slots[i + 1].id;
+            } else {
+                *next_slotid = CK_UNAVAILABLE_INFORMATION;
+            }
+        }
+    } else {
+        *next_slotid = CK_UNAVAILABLE_INFORMATION;
+    }
+
+    if (ret != CKR_OK) {
+        return ret;
+    }
+
+    /* acquire lock */
+    pthread_mutex_lock(&slot->sessions->lock);
+
+    if (slot->sessions->free_sessions) {
+        sess = slot->sessions->free_sessions;
+        LIST_REMOVE(slot->sessions->free_sessions, sess);
+        check_session = true;
+    } else {
+        sess = p11prov_session_new(provctx, slot);
+        if (sess == NULL) {
+            ret = CKR_HOST_MEMORY;
+            goto unlock;
+        }
+    }
+    if (sess->session == CK_INVALID_HANDLE) {
+        ret = p11prov_session_open(sess, false, NULL, 0);
```

Comment:
My concern is some form of deadlock if we do this, perhaps we should sleep for a "reasonable" time in case of session starvation and then still fail if a session does not come free in time.
The question is, what's a "reasonable" time?

Reply:
I am thinking we could wait but only for a "reasonable" time, what is reasonable TBD.
I do not want to risk deadlocks, where one operation holds the last free session and waits on another that can't get a session. Probably rare, but could happen. So we need a way to timeout eventually. In any case it makes no sense to wait indefinitely on a single thread application that just happens to have exhausted all its sessions (via a resource leak for example).